File: security-intel-hub/security_intel_hub/collectors/facebook.py
from __future__ import annotations

from datetime import datetime

from security_intel_hub.http import request_json
from security_intel_hub.models import IntelItem


class FacebookPageCollector:
    name = "facebook"

    def __init__(
        self,
        *,
        access_token: str,
        page_ids: list[str],
        graph_version: str = "v23.0",
        limit: int = 25,
        api_base: str = "https://graph.facebook.com",
    ) -> None:
        self.access_token = access_token
        self.page_ids = page_ids
        self.graph_version = graph_version.strip("/")
        self.limit = max(1, min(limit, 100))
        self.api_base = api_base.rstrip("/")

    def collect(self) -> list[IntelItem]:
        if not self.access_token or not self.page_ids:
            return []

        items: list[IntelItem] = []
        for page_id in self.page_ids:
            response = request_json(
                "GET",
                f"{self.api_base}/{self.graph_version}/{page_id}/posts",
                params={
                    "fields": "id,message,created_time,permalink_url,from",
                    "limit": self.limit,
                    "access_token": self.access_token,
                },
            )
            for post in response.get("data", []):
                message = post.get("message", "").strip()
                if not message:
                    continue
                author = (post.get("from") or {}).get("name", page_id)
                items.append(
                    IntelItem(
                        source=self.name,
                        source_id=post["id"],
                        title=_headline(message),
                        body=message,
                        url=post.get("permalink_url", ""),
                        author=author,
                        published_at=_parse_datetime(post.get("created_time")),
                        raw=post,
                    )
                )
        return items


def _headline(text: str, length: int = 96) -> str:
    text = " ".join(text.split())
    return text if len(text) <= length else f"{text[: length - 1]}..."


def _parse_datetime(value: str | None) -> datetime | None:
    if not value:
        return None
    return datetime.fromisoformat(value.replace("Z", "+00:00"))

File: backend/src/sources/facebook.ts
import type { FetchResult, ThreatIndicator } from '../types.js';
import { errorMessage, fetchWithTimeout } from '../util.js';
import { clampLimit, headline, parseIsoTime, scoreSocialText } from './social.js';

interface FacebookPost {
  id: string;
  message?: string;
  created_time?: string;
  permalink_url?: string;
  from?: { id?: string; name?: string };
}

interface FacebookPostsResponse {
  data?: FacebookPost[];
}

function parsePageIds(value: string | undefined): string[] {
  return (value ?? '')
    .split(',')
    .map((id) => id.trim())
    .filter(Boolean);
}

export async function fetchFacebookPages(): Promise<FetchResult<ThreatIndicator>> {
  const fetchedAt = Date.now();
  const accessToken = process.env.FACEBOOK_ACCESS_TOKEN?.trim();
  const pageIds = parsePageIds(process.env.FACEBOOK_PAGE_IDS);
  if (!accessToken || pageIds.length === 0) return { items: [], fetchedAt, error: null };

  try {
    const apiBase = (process.env.FACEBOOK_API_BASE?.trim() || 'https://graph.facebook.com').replace(
      /\/+$/,
      '',
    );
    const graphVersion = (process.env.FACEBOOK_GRAPH_VERSION?.trim() || 'v23.0').replace(
      /^\/+|\/+$/g,
      '',
    );
    const limit = clampLimit(process.env.FACEBOOK_LIMIT, 25, 1, 100);
    const items: ThreatIndicator[] = [];

    for (const pageId of pageIds) {
      const params = new URLSearchParams({
        fields: 'id,message,created_time,permalink_url,from',
        limit: String(limit),
        access_token: accessToken,
      });
      const res = await fetchWithTimeout(
        `${apiBase}/${graphVersion}/${encodeURIComponent(pageId)}/posts?${params.toString()}`,
        {},
        30_000,
      );
      if (!res.ok) throw new Error(`page ${pageId}: HTTP ${res.status}`);
      const data = (await res.json()) as FacebookPostsResponse;
      for (const post of data.data ?? []) {
        const message = post.message?.trim();
        if (!message) continue;
        const url = post.permalink_url || `https://facebook.com/${post.id}`;
        const scored = scoreSocialText(message);
        items.push({
          id: `facebook:${post.id}`,
          source: 'facebook',
          type: 'social_intel',
          indicator: url,
          indicatorType: 'url',
          severity: scored.severity,
          title: headline(message),
          description: message,
          tags: ['facebook', post.from?.name ?? pageId, ...scored.tags],
          reference: url,
          firstSeen: parseIsoTime(post.created_time),
        });
      }
    }

    return { items, fetchedAt, error: null };
  } catch (err) {
    return { items: [], fetchedAt, error: errorMessage(err) };
  }
}
